### DISCUSSION TITLE
Understanding graphs

### DISCUSSION THREAD

**Discussion by sameelab** [churchmanlab/genewalk#66] [General]
Hi  Churchman Lab!

I'm trying to understand the multi_graph.pkl file and the edge attributes.

For example, I ran genewalk for one gene (TP53) and I'm able to see this adjacency matrix

<img width="903" alt="image" src="https://user-images.githubusercontent.com/54965493/130835213-2ed67a35-592b-4c64-bc89-ed64ed84617e.png">

What does the attributes value between GO terms represent?

and

It's possible that the attribute value of the gene of interest to GO term is more than 1? 

Best 

Francisco Grisanti

**Comment by ri23** [comment]
Hi @sameelab 

Thanks for your interest in GeneWalk. 
multi_graph.pkl is a (pickled) networkx.MultiGraph object. It is very similar to a Python dictionary where essentially the keys are the network node identifiers and the values are a list of adjacent nodes. Details can be found here https://networkx.org/documentation/stable/reference/classes/multigraph.html.

Could you share the code how you generated the adjacency matrix starting with the networkx.MultiGraph object? 

I suspect that the numbers you mentioned are either a one hot encoding of the label attributes or the number of edges between the two nodes. Without the commands you ran I cannot determine this. 

Best,
Robert

**Reply by ri23** [reply]
Feel free to also check this thread for some code snippets to get started with the multi graph:
https://github.com/churchmanlab/genewalk/discussions/56#discussioncomment-632278

**Comment by sameelab** [comment]
Hi @ri23!

This was the code used to extract the adjacency matrix

`
### LOAD GRAPH
g = nx.read_gpickle('./test_tp53/multi_graph.pkl')

### PRUNE GRAPH IF NEEDED (LIST OF GENES)
genes_and_terms = ['TP53','GO:0000785','GO:0000976','GO:0000978','GO:0000981','GO:0000987','GO:0001046','GO:0001094','GO:0001216','GO:0001228']
g = g.subgraph(genes_and_terms)

### TO ADJ  
g_adj = nx.to_pandas_adjacency(g)
`

Thanks in advance!

**Comment by ri23** [comment]
Hi @sameelab 
Based on the documentation of nx.to_pandas_adjacency, it seems that the numbers in the adjacency matrix correspond to the number of edges (argument multigraph_weight defaults to sum) between two nodes in the multi_graph:
https://networkx.org/documentation/stable/reference/generated/networkx.convert_matrix.to_pandas_adjacency.html 
You can check this by calling a node pair in your subgraph g and check that the number corresponds to the number of edges.
For instance:
g['GO:0000978']['GO:0000976']
should return 4 edges.

For the GeneWalk algorithm these numbers actually don't matter because multiple edges in the multi graph are collapsed into 1 when sampling the random walks. So to be precise: if a neighbor is defined as an adjacent node connected with at least 1 edge in the multigraph, then the probability to walk to a neighbor is simply 1/n with n the number of neighbors for a node. 
You can see this in the code implementation here, where it simply samples a neighbor, irrespective of the number of edges between node and neighbor: https://github.com/churchmanlab/genewalk/blob/master/genewalk/deepwalk.py#L164

The rationale for this is that some genes and GO terms have multiple types of annotations, resulting in multiple edges between the same node pair. This is not really a biological aspect, because in reality there is really one physical interaction.

Hope this clarifies!
Best,
Robert
